### DISCUSSION TITLE
Feature request: merge SD.cpp and Microsoft CPU C++ ONNX Runtime adding ability to use ONNX face restoration models

### DISCUSSION THREAD

**Discussion by JohnClaw** [leejet/stable-diffusion.cpp#60] [Ideas]
Microsoft C++ ONNX Runtime for CPU: https://www.nuget.org/packages/Microsoft.ML.OnnxRuntime
ONNX models for face restoration: https://github.com/harisreedhar/Face-Upscalers-ONNX/releases/tag/Models

**Comment by JohnClaw** [comment]
This also may be useful: https://github.com/hpc203/CodeFormer-onnxrun-cpp-py
https://github.com/xmba15/onnx_runtime_cpp

**Comment by Green-Sky** [comment]
sd.cpp is based on ggml, not onnx. They are 2 different ml nn frameworks.

**Reply by JohnClaw** [reply]
Ok. Do you know ONNX cli.exe that can restore faces that were poorly rendered by SD.cpp?

**Reply by JohnClaw** [reply]
I know that ggml and onnx are different. But how does it matter when we are speaking about post-processing of an already generated image? For example, ggml_module.h of SD.cpp produces a cat.png. Then onnx_module.h of SD.cpp restores cat's face on that cat.png.
